File: src/services/syncService.js
/**
 * 云同步核心服务
 * 负责数据收集、推送、拉取、本地备份/恢复
 * HTTP 请求通过 Rust 后端代理（绕过 WebView2 网络限制）
 */
import { save, open } from '@tauri-apps/plugin-dialog'
import { writeTextFile, readTextFile } from '@tauri-apps/plugin-fs'
import { invoke } from '@tauri-apps/api/core'
import { syncState, getSupabaseConfig } from './syncState.js'
import i18n from '../i18n/index.js'

// ===== localStorage 需要同步的所有键 =====
const SYNC_KEYS = [
  'sn-notebooks',
  'sn-todolists',
  'sn-trash',
  'sn-item-trash',
  'sn-theme',
  'sn-accent',
  'sn-opacity',
  'sn-sidebar-visible',
  'sn-tabbar-visible',
  'sn-screenshot-hotkey',
  'sn-dev-mode',
  'sn-minimal-mode',
  'sn-active-book',
  'sn-active-todolist',
]

// ===== Rust HTTP 代理封装 =====

/** 安全提取错误消息（兼容 Tauri 各种错误对象结构） */
function safeErrorMessage(e) {
  if (typeof e === 'string') return e
  if (e?.message) return e.message
  try { return String(e) } catch { return i18n.global.t('common.unknownError') }
}

/**
 * 通过 Rust 后端发起 HTTP 请求
 * @param {string} url - 请求 URL
 * @param {string} method - HTTP 方法
 * @param {string} apiKey - Supabase anon key
 * @param {object|null} bodyObj - 请求体对象（可选）
 * @returns {Promise<{ ok: boolean, status: number, body: any }>}
 */
async function httpReq(url, method, apiKey, bodyObj = null, prefer = null) {
  const body = bodyObj ? JSON.stringify(bodyObj) : null
  console.log(`[sync] ${method} ${url}`)
  try {
    const result = await invoke('supabase_request', {
      url,
      method,
      apiKey,
      body,
      prefer,
    })
    console.log(`[sync] ${method} ${url} => ok=${result.ok} status=${result.status}`)
    return result
  } catch (e) {
    const err = safeErrorMessage(e)
    console.error(`[sync] ${method} ${url} 调用失败:`, err)
    throw new Error(err)
  }
}

// ===== 数据收集与恢复 =====

/**
 * 从 localStorage 收集所有需同步的数据
 * @returns {object} payload 对象
 */
export function collectLocalData() {
  const payload = {
    _meta: {
      version: 1,
      appVersion: '1.1.0',
      exportedAt: Date.now(),
    },
    data: {}
  }

  for (const key of SYNC_KEYS) {
    const val = localStorage.getItem(key)
    if (val !== null) {
      payload.data[key] = val
    }
  }

  return payload
}

/**
 * 将远程/导入的 payload 写回 localStorage
 * @param {object} payload - collectLocalData 生成的对象
 * @returns {{ success: boolean, message: string }}
 */
export function applyRemoteData(payload) {
  if (!payload || !payload.data) {
    return { success: false, message: i18n.global.t('sync.invalidFormat') }
  }

  const data = payload.data
  let count = 0

  for (const key of SYNC_KEYS) {
    if (key in data) {
      localStorage.setItem(key, data[key])
      count++
    }
  }

  // 派发事件通知所有组件重新加载数据
  window.dispatchEvent(new CustomEvent('sync-data-applied', {
    detail: { timestamp: Date.now() }
  }))

  return { success: true, message: i18n.global.t('sync.restoredItems', { count }) }
}

// ===== Supabase 推送 =====

/**
 * 推送本地数据到 Supabase 云端
 * @returns {Promise<{ success: boolean, message: string }>}
 */
export async function pushToCloud() {
  const config = getSupabaseConfig()
  if (!config.configured) {
    return { success: false, message: i18n.global.t('sync.notConfigured') }
  }

  syncState.status = 'syncing'
  syncState.errorMessage = ''

  try {
    const payload = collectLocalData()
    // 使用 POST + upsert：若 id=main 行不存在则自动创建，存在则更新
    const apiURL = `${config.url}/rest/v1/app_data`

    const res = await httpReq(apiURL, 'POST', config.key, {
      id: 'main',
      payload: payload,
    }, 'resolution=merge-duplicates,return=minimal')

    if (!res.ok) {
      const errBody = typeof res.body === 'object' ? JSON.stringify(res.body) : res.body
      throw new Error(`HTTP ${res.status}: ${errBody}`)
    }

    syncState.status = 'success'
    syncState.lastSyncTime = Date.now()
    localStorage.setItem('sn-last-sync-time', String(syncState.lastSyncTime))

    return { success: true, message: i18n.global.t('sync.pushSuccess') }
  } catch (e) {
    syncState.status = 'error'
    syncState.errorMessage = safeErrorMessage(e)
    return { success: false, message: i18n.global.t('sync.pushFailed', { error: safeErrorMessage(e) }) }
  }
}

// ===== Supabase 拉取 =====

/**
 * 从 Supabase 云端拉取数据
 * @returns {Promise<{ success: boolean, message: string }>}
 */
export async function pullFromCloud() {
  const config = getSupabaseConfig()
  if (!config.configured) {
    return { success: false, message: i18n.global.t('sync.notConfigured') }
  }

  syncState.status = 'syncing'
  syncState.errorMessage = ''

  try {
    const apiURL = `${config.url}/rest/v1/app_data?id=eq.main&select=payload,updated_at`
    const res = await httpReq(apiURL, 'GET', config.key)

    if (!res.ok) {
      const errBody = typeof res.body === 'object' ? JSON.stringify(res.body) : res.body
      throw new Error(`HTTP ${res.status}: ${errBody}`)
    }

    const rows = Array.isArray(res.body) ? res.body : []
    if (rows.length === 0) {
      syncState.status = 'idle'
      return { success: false, message: i18n.global.t('sync.noCloudData') }
    }

    const { payload, updated_at } = rows[0]
    if (!payload || !payload.data) {
      syncState.status = 'idle'
      return { success: false, message: i18n.global.t('sync.invalidCloudFormat') }
    }

    const result = applyRemoteData(payload)
    if (!result.success) return result

    syncState.status = 'success'
    syncState.lastSyncTime = Date.now()
    localStorage.setItem('sn-last-sync-time', String(syncState.lastSyncTime))

    const serverTime = updated_at ? new Date(updated_at).toLocaleString('zh-CN') : i18n.global.t('sync.unknownTime')
    return { success: true, message: i18n.global.t('sync.pullSuccess', { time: serverTime }) }
  } catch (e) {
    syncState.status = 'error'
    syncState.errorMessage = safeErrorMessage(e)
    return { success: false, message: i18n.global.t('sync.pullFailed', { error: safeErrorMessage(e) }) }
  }
}

// ===== 本地备份导出 =====

/**
 * 导出所有数据为 JSON 文件
 * @returns {Promise<{ success: boolean, message: string }>}
 */
export async function exportToJsonFile() {
  try {
    const payload = collectLocalData()
    const jsonStr = JSON.stringify(payload, null, 2)

    const now = new Date()
    const dateStr = `${now.getFullYear()}${String(now.getMonth()+1).padStart(2,'0')}${String(now.getDate()).padStart(2,'0')}_${String(now.getHours()).padStart(2,'0')}${String(now.getMinutes()).padStart(2,'0')}`
    const defaultName = `floatnote-backup-${dateStr}.json`

    const filePath = await save({
      title: i18n.global.t('sync.exportTitle'),
      defaultPath: defaultName,
      filters: [{ name: i18n.global.t('sync.jsonFilter'), extensions: ['json'] }],
    })

    if (!filePath) {
      return { success: false, message: i18n.global.t('common.cancelShort') }
    }

    await writeTextFile(filePath, jsonStr)
    return { success: true, message: i18n.global.t('sync.exportedTo', { path: filePath }) }
  } catch (e) {
    return { success: false, message: i18n.global.t('sync.exportFailed', { error: safeErrorMessage(e) }) }
  }
}

// ===== 本地备份导入 =====

/**
 * 从 JSON 文件导入数据
 * @returns {Promise<{ success: boolean, message: string }>}
 */
export async function importFromJsonFile() {
  try {
    const filePath = await open({
      title: i18n.global.t('sync.importTitle'),
      filters: [{ name: i18n.global.t('sync.jsonFilter'), extensions: ['json'] }],
      multiple: false,
    })

    if (!filePath) {
      return { success: false, message: i18n.global.t('common.cancelShort') }
    }

    const content = await readTextFile(filePath)
    let payload
    try {
      payload = JSON.parse(content)
    } catch {
      return { success: false, message: i18n.global.t('sync.invalidJson') }
    }

    // 校验数据格式
    if (!payload.data) {
      return { success: false, message: i18n.global.t('sync.invalidFormatNotBackup') }
    }

    // 确认覆盖
    const backupTime = payload._meta?.exportedAt ? new Date(payload._meta.exportedAt).toLocaleString('zh-CN') : i18n.global.t('sync.unknownTime')
    const dataCount = Object.keys(payload.data).length
    const confirmed = window.confirm(
      i18n.global.t('sync.importConfirmTitle', { time: backupTime, count: dataCount })
    )

    if (!confirmed) {
      return { success: false, message: i18n.global.t('common.cancelShort') }
    }

    const result = applyRemoteData(payload)
    return result
  } catch (e) {
    return { success: false, message: i18n.global.t('sync.importFailed', { error: safeErrorMessage(e) }) }
  }
}

// ===== 测试连接 =====

/**
 * 测试 Supabase 连接是否正常
 * @returns {Promise<{ success: boolean, message: string }>}
 */
export async function testConnection() {
  const config = getSupabaseConfig()
  if (!config.configured) {
    return { success: false, message: i18n.global.t('sync.notConfiguredShort') }
  }

  const testURL = `${config.url}/rest/v1/app_data?id=eq.main&select=id,updated_at`

  try {
    const res = await httpReq(testURL, 'GET', config.key)

    if (!res.ok) {
      const errBody = typeof res.body === 'object' ? JSON.stringify(res.body) : res.body
      if (res.status === 404 || (typeof errBody === 'string' && (errBody.includes('relation') || errBody.includes('does not exist')))) {
        return { success: false, message: i18n.global.t('sync.tableNotExist') }
      }
      if (res.status === 401 || res.status === 403) {
        return { success: false, message: i18n.global.t('sync.apiKeyInvalid') }
      }
      return { success: false, message: `HTTP ${res.status}: ${errBody}` }
    }

    const rows = Array.isArray(res.body) ? res.body : []

    if (rows.length === 0) {
      return { success: true, message: i18n.global.t('sync.connectionOk') }
    }

    const updated = rows[0].updated_at
      ? new Date(rows[0].updated_at).toLocaleString('zh-CN')
      : i18n.global.t('sync.noTime')
    return { success: true, message: i18n.global.t('sync.connectionOkWithTime', { time: updated }) }
  } catch (e) {
    return { success: false, message: i18n.global.t('sync.connectionFailed', { error: safeErrorMessage(e) }) }
  }
}

File: floatnote_android/lib/services/sync_service.dart
/// 云同步服务
/// 直连 Supabase REST API，与 PC 版 syncService.js 完全兼容

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/sync_payload.dart';
import 'storage_service.dart';

class SyncService {
  /// 需要同步的所有 localStorage 键（与 PC 版 SYNC_KEYS 一致）
  static const List<String> syncKeys = [
    'sn-notebooks',
    'sn-todolists',
    'sn-trash',
    'sn-item-trash',
    'sn-theme',
    'sn-accent',
    'sn-opacity',
    'sn-sidebar-visible',
    'sn-tabbar-visible',
    'sn-screenshot-hotkey',
    'sn-dev-mode',
    'sn-minimal-mode',
    'sn-active-book',
    'sn-active-todolist',
  ];

  final StorageService _storage;
  final http.Client _http;

  SyncService(this._storage, {http.Client? client})
      : _http = client ?? http.Client();

  // ===== 配置读取 =====

  String? get supabaseUrl => _storage.getString('sn-supabase-url');
  String? get supabaseKey => _storage.getString('sn-supabase-key');

  bool get isConfigured =>
      (supabaseUrl?.isNotEmpty ?? false) && (supabaseKey?.isNotEmpty ?? false);

  Future<void> saveConfig(String url, String key) async {
    await _storage.setString('sn-supabase-url', url.trim().replaceAll(RegExp(r'/+$'), ''));
    await _storage.setString('sn-supabase-key', key.trim());
  }

  // ===== 数据收集 =====

  SyncPayload collectLocalData() {
    final data = <String, String>{};
    for (final key in syncKeys) {
      final val = _storage.getString(key);
      if (val != null) data[key] = val;
    }
    return SyncPayload(
      meta: SyncMeta(exportedAt: DateTime.now().millisecondsSinceEpoch),
      data: data,
    );
  }

  /// 将远程数据写回本地
  int applyRemoteData(SyncPayload payload) {
    int count = 0;
    for (final key in syncKeys) {
      if (payload.data.containsKey(key)) {
        _storage.setString(key, payload.data[key]!);
        count++;
      }
    }
    return count;
  }

  // ===== Supabase HTTP 请求 =====

  Map<String, String> _headers({String? prefer}) {
    final h = <String, String>{
      'apikey': supabaseKey!,
      'Authorization': 'Bearer ${supabaseKey!}',
      'Content-Type': 'application/json',
    };
    if (prefer != null) h['Prefer'] = prefer;
    return h;
  }

  /// 推送到云端
  Future<SyncResult> pushToCloud() async {
    if (!isConfigured) {
      return const SyncResult(success: false, message: '请先配置 Supabase URL 和 API Key');
    }

    try {
      final payload = collectLocalData();
      final url = Uri.parse('${supabaseUrl!}/rest/v1/app_data');

      final response = await _http.post(
        url,
        headers: _headers(prefer: 'resolution=merge-duplicates,return=minimal'),
        body: jsonEncode({
          'id': 'main',
          'payload': payload.toJson(),
        }),
      ).timeout(const Duration(seconds: 15));

      if (response.statusCode < 300) {
        final now = DateTime.now().millisecondsSinceEpoch;
        await _storage.setString('sn-last-sync-time', now.toString());
        return const SyncResult(success: true, message: '推送成功');
      } else {
        return SyncResult(success: false, message: '推送失败: HTTP ${response.statusCode}\n${response.body}');
      }
    } catch (e) {
      return SyncResult(success: false, message: '推送失败: $e');
    }
  }

  /// 从云端拉取
  Future<SyncResult> pullFromCloud() async {
    if (!isConfigured) {
      return const SyncResult(success: false, message: '请先配置 Supabase URL 和 API Key');
    }

    try {
      final url = Uri.parse('${supabaseUrl!}/rest/v1/app_data?id=eq.main&select=payload,updated_at');

      final response = await _http.get(
        url,
        headers: _headers(),
      ).timeout(const Duration(seconds: 15));

      if (response.statusCode < 300) {
        final rows = jsonDecode(response.body) as List;
        if (rows.isEmpty) {
          return const SyncResult(success: false, message: '云端暂无数据，请先推送到云端');
        }

        final payloadJson = rows[0]['payload'] as Map<String, dynamic>?;
        if (payloadJson == null) {
          return const SyncResult(success: false, message: '云端数据格式无效');
        }

        final payload = SyncPayload.fromJson(payloadJson);
        final count = applyRemoteData(payload);

        final now = DateTime.now().millisecondsSinceEpoch;
        await _storage.setString('sn-last-sync-time', now.toString());

        final serverTime = rows[0]['updated_at']?.toString() ?? '未知';
        return SyncResult(success: true, message: '拉取成功 ($count 项)\n云端时间: $serverTime');
      } else {
        return SyncResult(success: false, message: '拉取失败: HTTP ${response.statusCode}\n${response.body}');
      }
    } catch (e) {
      return SyncResult(success: false, message: '拉取失败: $e');
    }
  }

  /// 测试连接
  Future<SyncResult> testConnection() async {
    if (!isConfigured) {
      return const SyncResult(success: false, message: '请先填写 URL 和 API Key');
    }

    try {
      final url = Uri.parse('${supabaseUrl!}/rest/v1/app_data?id=eq.main&select=id,updated_at');

      final response = await _http.get(
        url,
        headers: _headers(),
      ).timeout(const Duration(seconds: 15));

      if (response.statusCode == 401 || response.statusCode == 403) {
        return const SyncResult(success: false, message: 'API Key 无效或权限不足');
      }

      if (response.statusCode == 404) {
        return const SyncResult(
            success: false,
            message: 'app_data 表不存在\n请在 Supabase SQL Editor 中建表');
      }

      if (response.statusCode < 300) {
        final rows = jsonDecode(response.body) as List;
        if (rows.isEmpty) {
          return const SyncResult(success: true, message: '连接正常！\n尚未推送过数据');
        }
        final updated = rows[0]['updated_at']?.toString() ?? '无';
        return SyncResult(success: true, message: '连接正常！\n上次更新: $updated');
      }

      return SyncResult(success: false, message: 'HTTP ${response.statusCode}');
    } catch (e) {
      return SyncResult(success: false, message: '连接失败: $e');
    }
  }

  // ===== 本地备份 =====

  /// 导出为 JSON 字符串
  String exportToJson() {
    final payload = collectLocalData();
    return const JsonEncoder.withIndent('  ').convert(payload.toJson());
  }

  /// 从 JSON 字符串导入
  SyncResult importFromJson(String jsonStr) {
    try {
      final parsed = jsonDecode(jsonStr) as Map<String, dynamic>;
      if (!parsed.containsKey('data')) {
        return const SyncResult(success: false, message: '数据格式无效：缺少 data 字段');
      }
      final payload = SyncPayload.fromJson(parsed);
      final count = applyRemoteData(payload);
      return SyncResult(success: true, message: '导入成功 ($count 项)');
    } catch (e) {
      return SyncResult(success: false, message: '导入失败: $e');
    }
  }

  void dispose() {
    _http.close();
  }
}
